geowrangler/datasets/ookla.py

```diff
@@ -58,13 +279,189 @@ def download_ookla_file(
+    )
+    return download_ookla_file(
+        type_=type_,
+        year=year,
+        quarter=quarter,
+        directory=directory,
+    )
+
+
+# Cell
+
+
+def download_ookla_parallel(num_expected_ookla_files, type_, year, directory):
+    items = [
+        (str(i), type_, year, directory)
+        for i in range(1, num_expected_ookla_files + 1, 1)
+    ]
+    parallel(parallel_download, items, threadpool=True, progress=True)
+
+
+# Cell
+
+
+def download_ookla_year_data(type_, year, cache_dir, use_cache=True):
+
+    "Download ookla data for a specifed type (fixed or mobile) and year. Data for all 4 quarters will be downloaded."
+
+    # Determine number of expected data for type_ and year, specified by OoklaFile(type, year, quarter)
+    available_ookla_files = list_ookla_files()
+    expected_ookla_files = {}
+    for quarter in [1, 2, 3, 4]:
+        quarter_ookla_file = OoklaFile(str(type_), str(year), str(quarter))
+        if quarter_ookla_file in available_ookla_files.keys():
+            expected_ookla_files[quarter_ookla_file] = available_ookla_files[
+                quarter_ookla_file
+            ]
+    num_expected_ookla_files = len(expected_ookla_files)
+
+    if num_expected_ookla_files == 0:
+        logger.warning(f"Ookla data: No data available for {type_} and {year}")
+    else:
+        logger.info(
+            f"Ookla Data: Number of available files for {type_} and {year}: {num_expected_ookla_files}"
+        )
+
+    ookla_cache_dir = os.path.join(cache_dir, "ookla/")
+    type_year_cache_dir = os.path.join(ookla_cache_dir, type_, str(year))
+
+    # Check if the cached data is valid. Otherwise, we have to re-download.
+    # For Ookla, we need to check if we've downloaded all expected files for that year.
+    cached_data_available = (
+        os.path.exists(type_year_cache_dir)
+        and len(os.listdir(type_year_cache_dir)) == num_expected_ookla_files
+    )
+
+    logger.info(
+        f"Ookla Data: Cached data available for {type_} and {year} at {type_year_cache_dir}? {cached_data_available}"
+    )
+
+    # Download if cache is invalid or user specified use_cache = False
+    if not cached_data_available or not use_cache:
+        logger.info(
+            f"Ookla Data: Re-initializing Ookla type/year cache dir at {type_year_cache_dir}..."
+        )
+        # Re-create the country cache dir and start over to fix any corrupted states
+        shutil.rmtree(type_year_cache_dir, ignore_errors=True)
+        Path(type_year_cache_dir).mkdir(parents=True, exist_ok=True)
+
+        # This downloads a parquet file to the type_year_dir for each quarter
+        download_ookla_parallel(
+            num_expected_ookla_files, type_, year, type_year_cache_dir
+        )
+
+        logger.info(
+            f"Ookla Data: Successfully downloaded and cached Ookla data for {type_} and {year} at {type_year_cache_dir}!"
+        )
+
+    return type_year_cache_dir
+
+
+# Cell
+
+
+def add_ookla_features(
```

Comment:
It might be best to move this level of processing into a separate dir like `geowrangler/processing/` 

I envisioned dataset module to be purely for fetching datasets.

Reply:
Hello, just wanted to chime in as well. I think this organization is ok. Bundling all dataset-related features together actually makes it easier in my opinion. Like if I want to know everything that has to do with getting data and generating features from Ookla, it's all in one place (which is the typical use-case). In other words, I think it's quite rare I'd use the Ookla-related functions separately from each other. There's not much use for pure download without any feature engineering. And I cannot do feature engineering without downloading the data. This specific feature engineering util is not general anyway; it's really just dataset-specific. 

So a dir like `geowrangler/processing/` makes sense for me for the general functions (like vector zonal stats, area zonal stats, etc), but not this where it's specifically for Ookla.

Reply:
I'm thinking about separation of concerns and possible regressions if we keep them in the same place. I've seen loading of data as a separate tasks from feature eng.

Reply:
I think there can be a balance here. 

The most reusable feature engineering functions from ookla have to do with filtering the data within the aois and then aggregating them by year and caching the results (this is in addition to caching the parquet files themselves). This speeds up your per-country/per-region workflows because  the annual averaged ookla data for your aoi is cached (while avoiding having to average the annual data on a global scale). 

I've separated the ookla processing specific parts (e.g. aggregation area zonal stats  in the `add_ookla_features` methods which can be moved back to either poverty mapping or spun out to a separate geowrangler ookla processing module. I plan to retain the annual averaging part in the geowrangler.datasets.ookla module since the results are cached and is part of the OoklaDataManager cached retrieval functionality